test/tests/functional/pbs_hook_execjob_abort.py

```diff
@@ -0,0 +1,282 @@
+#
+# PBS Pro is free software. You can redistribute it and/or modify it under the
+# terms of the GNU Affero General Public License as published by the Free
+# Software Foundation, either version 3 of the License, or (at your option) any
+# later version.
+#
+# PBS Pro is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.
+# See the GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+# Commercial License Information:
+#
+# For a copy of the commercial license terms and conditions,
+# go to: (http://www.pbspro.com/UserArea/agreement.html)
+# or contact the Altair Legal Department.
+#
+# Altair’s dual-license business model allows companies, individuals, and
+# organizations to create proprietary derivative works of PBS Pro and
+# distribute them - whether embedded or bundled with other software -
+# under a commercial license agreement.
+#
+# Use of Altair’s trademarks, including but not limited to "PBS™",
+# "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+# trademark licensing policies.
+
+import time
+from tests.functional import *
+from ptl.utils.pbs_logutils import PBSLogUtils
+
+
+class TestPbsExecjobAbort(TestFunctional):
+    """
+    Tests the hook event execjob_abort for when a job prematurely exits
+    during startup and the epilogue hook or end hook may not always execute.
+    """
+    logutils = PBSLogUtils()
+
+    def setUp(self):
+        if len(self.moms) != 3:
+            self.skipTest('test requires three MoMs as input, ' +
+                          'use -p moms=<mom1>:<mom2>:<mom3>')
+        TestFunctional.setUp(self)
+        self.momC = self.moms.values()[2]
+
+        # execjob_abort hook
+        self.abort_hook_body = """import pbs
+e=pbs.event()
+pbs.logmsg(pbs.LOG_DEBUG, "called execjob_abort hook")
+
+def print_attribs(pbs_obj, header):
+    for a in pbs_obj.attributes:
+        v = getattr(pbs_obj, a)
+        if v and str(v) != "":
+            pbs.logmsg(pbs.LOG_DEBUG, "%s: %s = %s" % (header, a, v))
+print_attribs(e.job, "JOB")
+
+for vn in e.vnode_list:
+    v = e.vnode_list[vn]
+    print_attribs(v, "vnode_list[" + vn + "]")
+"""
+        # instantiate execjob_abort hook
+        hook_event = 'execjob_abort'
+        hook_name = 'abort'
+        a = {'event': hook_event, 'enabled': 'true'}
+        self.server.create_import_hook(hook_name, a, self.abort_hook_body)
+
+        # execjob_abort hook with sleep
+        self.abort1_hook_body = """import pbs
+import time
+e=pbs.event()
+pbs.logmsg(pbs.LOG_DEBUG, "called execjob_abort hook")
+time.sleep(2)
+"""
+        # execjob_prologue hook, unhandled exception
+        self.prolo_hook_body = """import pbs
+e=pbs.event()
+job=e.job
+if %s job.in_ms_mom():
+    x
```

Comment:
'x' will cause an unhandled exception and the hook will reject.

Reply:
It might be better to use something more self explanatory like...
```
raise AssertionError
```

Reply:
I could also do an explicit reject: e.reject("hook rejected") instead of raising an exception, but the test is to have an unhandled exception that will cause a hook reject.

Reply:
Or you could raise an exception using Python syntax like:
```
raise AssertionError
```

Reply:
Done